api/filters/fsslice/fieldspec.go

```diff
@@ -0,0 +1,140 @@
+// Copyright 2019 The Kubernetes Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package fsslice
+
+import (
+	"strings"
+
+	"sigs.k8s.io/kustomize/api/types"
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+// fieldSpecFilter applies a single fieldSpec to a single object
+// fieldSpecFilter stores internal state and should not be reused
+type fieldSpecFilter struct {
+	// FieldSpec contains the path to the value to set.
+	FieldSpec types.FieldSpec `yaml:"fieldSpec"`
+
+	// Set the field using this function
+	SetValue SetFn
+
+	// CreateKind defines the type of node to create if the field is not found
+	CreateKind yaml.Kind
+
+	// path keeps internal state about the current path
+	path []string
+}
+
+func (fltr fieldSpecFilter) Filter(obj *yaml.RNode) (*yaml.RNode, error) {
+	// check if the FieldSpec applies to the object
+	if match, err := isMatchGVK(fltr.FieldSpec, obj); !match || err != nil {
+		return obj, err
+	}
+	fltr.path = strings.Split(fltr.FieldSpec.Path, "/")
+	if err := fltr.filter(obj); err != nil {
+		return nil, err
+	}
+	return obj, nil
+}
+
+func (fltr fieldSpecFilter) filter(obj *yaml.RNode) error {
+	if len(fltr.path) == 0 {
+		// found the field -- set its value
+		return fltr.SetValue(obj)
+	}
+	switch obj.YNode().Kind {
+	case yaml.SequenceNode:
+		return fltr.seq(obj)
+	case yaml.MappingNode:
+		return fltr.field(obj)
+	}
+	// not found -- this might be an error since the type doesn't match
+	return nil
+}
+
+// field calls filter on the field matching the next path element
+func (fltr fieldSpecFilter) field(obj *yaml.RNode) error {
+	fieldName, isSeq := isSequenceField(fltr.path[0])
+
+	// lookup the field matching the next path element
+	var lookupField yaml.Filter
+	switch {
+	case !fltr.FieldSpec.CreateIfNotPresent || fltr.CreateKind == 0 || isSeq:
+		// dont' create the field if we don't find it
+		lookupField = yaml.Lookup(fieldName)
+	case len(fltr.path) <= 1:
+		// create the field if it is missing: use the provided node kind
+		lookupField = yaml.LookupCreate(fltr.CreateKind, fieldName)
+	default:
+		// create the field if it is missing: must be a mapping node
+		lookupField = yaml.LookupCreate(yaml.MappingNode, fieldName)
+	}
+
+	// locate (or maybe create) the field
+	field, err := obj.Pipe(lookupField)
+	if err != nil || field == nil {
+		return err
```

Comment:
due to recursion, please wrap err with more context.

return error.Wrap(err, "node=%v, lookupField=%s", ...)

Reply:
added the fieldName, the other parts will be added by at the entry point.